tests/xharness/XmlResultParser.cs

```diff
@@ -0,0 +1,257 @@
+﻿using System;
+using System.IO;
+using System.Xml;
+
+namespace xharness {
+	public static class XmlResultParser {
+
+		public enum Jargon {
+			TouchUnit,
+			NUnit,
+			xUnit,
+			Text,
+		}
+
+		// test if the file is valid xml, or at least, that can be read it.
+		public static bool IsValidXml (string path, out Jargon type)
+		{
+			type = Jargon.Text;
+			if (!File.Exists (path))
+				return false;
+
+			bool isXml = false;
+			using (var stream = File.OpenText (path)) {
+				string line;
+				while ((line = stream.ReadLine ()) != null) { // special case when get got the tcp connection
+					if (line.Contains ("ping"))
+						continue;
+					if (line.Contains ("TouchUnitTestRun")) {
+						type = Jargon.TouchUnit;
+						isXml = true;
+						break;
+					}
+					if (line.Contains ("nunit-version")) {
+						type = Jargon.NUnit;
+						isXml = true;
+					}
+					if (line.Contains ("xUnit")) {
+						type = Jargon.xUnit;
+						isXml = true;
+						break;
+					}
+				}
+			}
+			return isXml;
+		}
+
+		static (string resultLine, bool failed) ParseTouchUnitXml (StreamReader stream, StreamWriter writer)
+		{
+			long total, errors, failed, notRun, inconclusive, ignored, skipped, invalid;
+			total = errors = failed = notRun = inconclusive = ignored = skipped = invalid = 0L;
+
+			using (var reader = XmlReader.Create (stream)) {
+				while (reader.Read ()) {
+					if (reader.NodeType == XmlNodeType.Element && reader.Name == "test-results") {
+						total = long.Parse (reader ["total"]);
+						errors = long.Parse (reader ["errors"]);
+						failed = long.Parse (reader ["failures"]);
+						notRun = long.Parse (reader ["not-run"]);
+						inconclusive = long.Parse (reader ["inconclusive"]);
+						ignored = long.Parse (reader ["ignored"]);
+						skipped = long.Parse (reader ["skipped"]);
+						invalid = long.Parse (reader ["invalid"]);
```

Comment:
potential for exceptions, `TryParse` would be safer
I guess it depends on how much do you trust the XML input ;-)